### PR TITLE
Gossip no longer pushes/pulls from nodes with a different shred version

### LINKED ISSUE #9855: RPC: Get recent blockhash fails on devnet
#### Problem
```json
{"method":"getRecentBlockhash","jsonrpc":"2.0","params":[],"id":"c159a4fb-99a0-4c77-b251-edfc93e5d83a"}
```

Received the following error when fetching the recent blockhash from https://devnet.solana.com
```
Failed to update cluster Error: failed to get recent blockhash: Error: failed to get recent blockhash: Cluster largest_confirmed_root 5540360 does not exist on node. Node root: 5540392
```

#### Proposed Solution
TBD

tag @CriesofCarrots

### PR DESCRIPTION
**Pull request by mvines**
Fixes #9855

### CODE REVIEW

File: core/src/cluster_info.rs [hunk 1]
```diff
@@ -361,51 +365,59 @@ impl ClusterInfo {
         let now = timestamp();
         let mut spy_nodes = 0;
         let mut archivers = 0;
+        let mut different_shred_nodes = 0;
```
Comment (mvines) [inline core/src/cluster_info.rs:368]: `contact_info_trace()` changes filters out nodes of different shreds for more concise view of the real network

File: core/src/crds_gossip_pull.rs [hunk 2]
```diff
@@ -165,13 +166,20 @@ impl CrdsGossipPull {
...
         crds.table
             .values()
             .filter_map(|v| v.value.contact_info())
-            .filter(|v| v.id != *self_id && ContactInfo::is_valid_address(&v.gossip))
+            .filter(|v| {
+                v.id != *self_id
+                    && ContactInfo::is_valid_address(&v.gossip)
+                    && (self_shred_version == 0
+                        || v.shred_version == 0
+                        || self_shred_version == v.shred_version)
```
Comment (mvines) [inline core/src/crds_gossip_pull.rs:181]: The real change, part I

File: core/src/crds_gossip_push.rs [hunk 3]
```diff
@@ -288,13 +289,20 @@ impl CrdsGossipPush {
...
             .values()
             .filter(|v| v.value.contact_info().is_some())
             .map(|v| (v.value.contact_info().unwrap(), v))
-            .filter(|(info, _)| info.id != *self_id && ContactInfo::is_valid_address(&info.gossip))
+            .filter(|(info, _)| {
+                info.id != *self_id
+                    && ContactInfo::is_valid_address(&info.gossip)
+                    && (self_shred_version == 0
+                        || info.shred_version == 0
+                        || self_shred_version == info.shred_version)
```
Comment (mvines) [inline core/src/crds_gossip_push.rs:304]: The real change, part II

**Comment by mvines** [comment]
@sakridge - ping

**Comment by sakridge** [comment]
Not a blocker for this, but I wonder if spy nodes should also just adopt a version and then we don't need the special version==0 case. Is there something fundamental that prevents that?

**Comment by mvines** [comment]
> Not a blocker for this, but I wonder if spy nodes should also just adopt a version and then we don't need the special version==0 case. Is there something fundamental that prevents that?

Nothing fundamental that I can think of.  I find some utility in being able to view all the nodes in gossip with `solana-gossip spy`, but that's not that important.   But if the special version was dropped, we'd probably then need to add a `--shred-version` argument to `solana-gossip spy`.  Also not a big deal.   I guess I don't have very strong feelings about it either way

**Pull request merged**